teps/0048-pipeline-as-code.md

```diff
@@ -0,0 +1,214 @@
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+
+## Summary
+
+The pipelines as code technique emphasizes that the configuration of delivering
+pipelines that build, test and deploy our applications or infrastructure should
+be treated as code; they should be placed under source control and modularized
+in reusable components with automated testing and deployment.
+
+
+## Motivation
+
+The Tekton pipeline definitions are currently dissociated from a source code repository. 
+They may be located in the same repository as they code they should build, test and deploy,
+but Tekton is not aware of them. Tekton expects the pipeline definition to be deployed to a
+kubernetes cluster. Even when the Tekton resources are deployed to the cluster from the
+Repository, it's not easy to test changes to them, as their definition won't be updated until
+they are redeployed to the cluster. 
+
+The reason for the current behaviour is that Tekton is non-prescriptive about how to set up
+CI and CD pipelines. It aims to provide the building blocks for creating such pipelines and it leaves
+it up to the user to decide how to integrate the pipeline in the overall workflow. 
+
+The use case of a single repo, running CI/CD jobs in isolation, is very common and
+we wish to provide dedicated support for it, by providing an easy way for users to
+run pipelines and tasks stored in a specific folder of the repo. The same tool can be used by
+pipeline and task authors for quick testing and experimentation of their code.
+
+### Goals
+
+A way for users to add a tekton pipeline in a repository and getting it
+automatically picked up for CI or CD.
+
+> The solution should be as much as possible git hosting system agnostic.
+
+### Non-Goals
+
+- GitOps: Pipelines as Code does not take care of the [GitOps primitives](https://www.weave.works/technologies/gitops/), it
+  simply applies tekton pipelines that may take care of some GitOps operation.
+
+### Use Cases
+
+- A User have a source code repository based on some Golang backend.
+- User add a webhook (or other means) from its repository to pipelines as a code.
+- The user add a directory called in `.tekton` with some pipeline definition for
+  testing its repository.
+- User sending a PR with its change.
+- The source code is automatically tested with the pipeline definition in `.tekton`.
+
+Subsequently if the user needs to add another component to its source code (i.e:
+a javascript based frontend) and need to make a changes to the pipeline
+definition to do the testing of that new component, the change to the pipeline
+would only apply for that Pull Request until it will be merged.
```

Comment:
As for GitHub actions, once the `.github/workflows` yml file is exists, a new pipeline will be triggered when a Pull Request is created but not merged, I wonder if we should do the same design.

Reply:
@FogDong could you explain a bit more? what you are describing sounds similar to what @chmouel described above but I think I'm missing some details

Reply:
I re-read it and realized that I had misunderstood it 😅